src/commands.cpp

```diff
@@ -2003,60 +2017,211 @@ void CommandGetUA::procresult()
 
         if (!(ptr = client->json.getvalue()) || !(end = strchr(ptr, '"')))
         {
-            return(client->app->getua_result(API_EINTERNAL));
+            client->app->getua_result(API_EINTERNAL);
+            return;
         }
 
         // if there's no avatar, the value is "none" (not Base64 encoded)
         if (attributename == "+a" && !strncmp(ptr, "none", 4))
         {
-            return(client->app->getua_result(API_ENOENT));
+            client->app->getua_result(API_ENOENT);
+            return;
         }
 
-        int l;
+        // convert from ASCII to binary the received data
+        unsigned int datalen;
         byte* data;
-        l = (end - ptr) / 4 * 3 + 3;
-        data = new byte[l];
-        l = Base64::atob(ptr, data, l);
-
-        if (attributename == "*!lstint" || attributename == "*!authring")
+        datalen = (end - ptr) / 4 * 3 + 3;
+        data = new byte[datalen];
+        datalen = Base64::atob(ptr, data, datalen);
+
+        bool signkeyUpdated = false;
+        bool chatkeyUpdated = false;
+        bool keyringSet = (client->signkey && client->chatkey);
+        string tmpstr;
+//        bool nonHistoric = (attributename.at(1) == '!');
+
+        // handle the attribute data depending on the scope
+        char scope = attributename.at(0);
+        switch (scope)
         {
-            string d;
-            d.assign((char*)data, l);
+        case '*':   // private
 
-            // Is the data a multiple of the cipher blocksize, then we're using
-            // a zero IV.
-            if (l % client->key.BLOCKSIZE == 0)
+            // decrypt the data and build the TLV records
+            tlvRecords = TLVstore::containerToTLVrecords(string((char*) data, datalen), &client->key);
+            if (!tlvRecords)
             {
-                if (!PaddedCBC::decrypt(&d, &client->key))
+                LOG_err << "Cannot extract TLV records for private attribute " << attributename;
+                client->app->getua_result(API_EINTERNAL);
+                delete [] data;
+                return;
+            }
+
+            if (attributename == "*keyring")
```

Comment:
I think it might be a good idea to extract processing for certain private and public attributes that require special treatment (keyring, pub keys, signatures) for sanity checks/"healing" to an external function to not inflate and overcomplicate this function.
But that's more an issue for further refactoring, the procedure here seems good in general.